### PR TITLE
onnx full passing

### PR DESCRIPTION
**Pull request by geohotstan**
Failing tests (53): https://gist.github.com/geohotstan/db8d722fe311db2b06709eb2a5276a00
Todo:
- get to 0 failing
- clean up code (lots of clean up) 
- write better implementations for ops (Slice, gather, resize, pad, ...) and type parse. I'm using too many kernels (openpilot test failing since 205 (<199) kernels used), and some of my current implementations are admittedly, quite ugly. Tinygrad is  elegant.
- maybe also benchmarking (onnx runtime??? I have no idea how to do this)

Posting this draft PR since I've been perma stuck on ~~linear~~ and cubic interpolation for Resize for awhile now and also kinda unsure if the code I'm writing is totally disgusting or passable.... 😢
Not sure if anyone else have any ideas on interpolation.
For bounty hunters, feel free to take parts for your onnx fixes or write an implementation better than mine so that you can help me learn and we can share the fun!

I'm still dead set on getting to 0 failing though.

### CODE REVIEW

**Comment by geohot** [comment]
Up to 10 failing (on each platform) is acceptable for the bounty

**Comment by tinyb0t** [comment]
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/tensor.py                         1               1
------------------------------------------------------------
lines added in the tinygrad folder: 0
```

**Comment by geohot** [comment]
Careful, I think openpilot test failure is real

**Comment by geohotstan** [comment]
> Careful, I think openpilot test failure is real

Yeah you're right, Test openpilot model correctness (float32) also failed with mismatch elements, crap.
Will investigate.

**Comment by geohotstan** [comment]
https://github.com/geohot/tinygrad/pull/1101 introduced 20 more failed tests aside from `backend_test.exclude('float64')`

Tests:
- `test_reduce_log_sum_exp*`
- `test_operator_add*`

Will exclude them for now..?

**Comment by geohot** [comment]
Ugh really? Is float64 actually needed for those tests?

**Comment by geohotstan** [comment]
> Ugh really? Is float64 actually needed for those tests?

yeah 😞 I think so.
For those tests, both the inputs (`numpy.dtype[float64]`) and the ref_output (`DESIRED: dtype('float64')`) expects float64

**Comment by geohot** [comment]
I'm fine with skipping the float64 ones for now, just mark that section of skips.

**Comment by geohot** [comment]
Still seeing `/home/runner/work/_temp/4a9f8191-c275-4935-9d41-8db11820c215.sh: line 1:  3123 Segmentation fault      (core dumped) ALLOWED_KERNEL_COUNT=199 FLOAT16=1 DEBUGCL=1 GPU=1 IMAGE=2 python3 openpilot/compile.py`

What are the 29 failures left?

**Comment by geohotstan** [comment]
running `FLOAT16=1 DEBUGCL=1 GPU=1 IMAGE=2 python3 openpilot/compile.py` locally without the `ALLOWED_KERNEL_COUNT` flag it seems to pass with 
```
kernel count: 206
...
total runtime: 14.83 ms   
wall time: 16.87 ms
network using 1.50 GOPS with runtime 14.83 ms that's 101.01 GFLOPS
```




> What are the 29 failures left?

29 failures: https://gist.github.com/geohotstan/f5284c17d1feb8cb39ddbd18f71a2b86

**Comment by geohot** [comment]
I mean...it has to pass in CI, it's reliable in master. So there must be some regression.

Cool those failures seem reasonable to work on, nothing crazy that's like not a real thing.

**Comment by geohotstan** [comment]
Temporary fix to get it passing in CI.
Change: `ALLOWED_KERNEL_COUNT=199` -> `ALLOWED_KERNEL_COUNT=207`

I'll try to decrease the number or kernels used... The additional kernels used is mainly introduced by the new Gather op, so I'll try my hardest in finding a more efficient implementation asap.

I will revert this change once I find a better way of doing Gather.

**Comment by geohot** [comment]
Allowed kernels definitely can't change...is resize stuff hard? Might be okay with skipping that.

**Comment by geohotstan** [comment]
> Allowed kernels definitely can't change...is resize stuff hard? Might be okay with skipping that.

I have reverted the allowed kernels change. It's back to 199 and CI still passes. 😸

I think I can complete the resize stuff (it's a mess with lot of lines of code unfortunately... lots of modes and options). 
Might have to hand code the math for cubic interpolation also...
I'll give resize another try.
Good news is once resize is done, it should be close to 10 test failures with TORCH=1.

was away from me computer past few days, almost back. Will push for finish line.

**Comment by geohot** [comment]
We ready to get this merged? Just skip the last 9 under "skipped less than 10"

**Comment by geohotstan** [comment]
> We ready to get this merged? Just skip the last 9 under "skipped less than 10", clean it up, and let's upstream this!

Still need to clean things up a bit :s
Very soon!

**Comment by geohotstan** [comment]
Final few TODOs:

- `test_squeezenet` model test is failing on TORCH (also failing on master) due to a pretty deep bug.. (passes on CPU)
- implement SoftmaxCrossEntropyLoss
- final few clean ups

should I make sure tests pass for all backends too? Is that what is meant for "(on each platform)"?

**Comment by geohot** [comment]
Yea, on each platform means the different backends.

If you find bugs in the backends, feel free to file issues and it's likely someone else will fix them.

File: tinygrad/runtime/ops_cpu.py [hunk 1]
```diff
@@ -45,6 +44,8 @@ def mulacc(a, b, new_shape):
 class RawNumpyBuffer(RawBuffer):
   def __init__(self, size:int, dtype:DType, buf:Optional[np.ndarray]=None): super().__init__(size, dtype, buf if buf is not None else np.empty([size], dtype.np))
   @classmethod
-  def fromCPU(cls, x): return cls(x.size, dtypes.from_np(x.dtype), x)
+  def fromCPU(cls, x):
```
Comment (geohot) [inline tinygrad/runtime/ops_cpu.py:47]: This should not be here. I think the promote types might be a bug too. Can you submit this is as a separate PR?

**Comment by geohotstan** [comment]
Are we sure that parsing the shape from `type_proto` and then comparing it to the tensor after initiated in `input_tensor` is necessary?
There are some really weird type_protos
such as:
`bert` in `external_model_benchmark`
the first Tensor that comes in is `<Tensor <LB (1,)...>`
but it's `type_proto` is 
```
tensor_type {
  elem_type: 7
  shape {
    dim {
      dim_param: "unk__492"
    }
  }
}
```
There is no `dim_value` here so shape cannot be determined by `type_proto` I don't think.
and so `assert input_shape == shape, f"wrong shape for input {inp.name}, {input_shape} isn't {shape}"` fails

I've managed to fix the `type_proto` related problems for external_test_onnx_backend though.

**Comment by geohotstan** [comment]
> can you update the falling tests?

on TORCH, CPU backend: https://gist.github.com/geohotstan/8b9dc63081f259b7efc4a290b5cd07ad
on GPU, METAL, MPS: tests that involve 0 shape tensors and really weird inaccuracy for mish op 
on Compiled backends: also 0 shape tests failing because linearizer doesn't allow 0 shape reshape
on LLVM: https://github.com/tinygrad/tinygrad/issues/1479

**Comment by geohot** [comment]
Want to file issues for the 0 shape failures with a minimal case? That sounds like a bug

**Comment by ZenginU** [comment]
Hello @geohotstan ,

What do you think about "# unsupported (strange) ops", I have implemented "erf" not much but small improvement if you are not going to try them maybe I can take a look at it as well.

**Comment by geohotstan** [comment]
> Hello @geohotstan ,
> 
> What do you think about "# unsupported (strange) ops", I have implemented "erf" not much but small improvement if you are not going to try them maybe I can take a look at it as well.

Yeah go right ahead, any work is much appreciated :D 
btw what's erf?

File: extra/onnx.py [hunk 2]
```diff
@@ -114,7 +116,7 @@ def run_onnx(inputs={}, debug=False):
       if inp.name in tensors: continue
       shape = type_parse(inp.type)
       # this works for gpt3 but does not work elsewhere
-      # if len(shape) >= 1 and shape[0] == 0 and shape != (0,): shape = tuple([1]+list(shape[1:]))   # 1 batch size
+      if len(shape) >= 1 and shape[0] == 0 and shape != (0,): shape = tuple([1]+list(shape[1:]))   # 1 batch size
```
Comment (gijskoning) [inline extra/onnx.py:119]: Simpler and same behaviour 
```suggestion
      if len(shape) > 1 and shape[0] == 0: shape = (1,)+shape[1:]   # 1 batch size
```

Comment (geohotstan) [reply extra/onnx.py]: I don't think reshaping 0 to 1 works... (I may be wrong) 
There are a lot of other type_proto that have really weird structures and those fail when reshaping like this... 

What I have now is:
optional_type has element_type = tensor_type or sequence_type
and sequence_type has element_type = tensor_type

so when tensor_type does not have dim_value, I assume it's a variable tensor and skip the shape assert:
`assert input_shape == shape, f"wrong shape for input {inp.name}, {input_shape} isn't {shape}"`
(since checking if shape is equal when type_proto doesn't provide a shape is probably not necessary)

**Comment by geohot** [comment]
Is this ever going to be ready to merge?

**Comment by geohotstan** [comment]
currently:

- yolov8-onnx.py runs
- openpilot from external_model_benchmark does not have numerical error mentioned in #1681 
- I skipped more tests for specific backends related to some bugs still WIP (documented)
- for all backends except CUDA (I can't test this) there's 0 failing tests

for CPU=1 for example: `=============== 916 passed, 1718 skipped, 68 warnings in 16.23s ================`
with `./run_multibackend.sh test/external/external_test_onnx_backend.py`
```
----------------------------------------------------------------------                                                                           
CPU
Ran 2634 tests in 14.652s                                                                                                                                                                                     
OK (skipped=1718)     
----------------------------------------------------------------------
GPU
Ran 2634 tests in 7.247s
OK (skipped=1721)
----------------------------------------------------------------------
METAL
Ran 2634 tests in 5.252s
OK (skipped=1724)
----------------------------------------------------------------------
CLANG
Ran 2634 tests in 19.727s
OK (skipped=1718)
----------------------------------------------------------------------
LLVM
Ran 2634 tests in 6.725s
OK (skipped=1718)
----------------------------------------------------------------------
TORCH
Ran 2634 tests in 12.190s
OK (skipped=1746)
```

**Comment by geohot** [comment]
Going to merge, bounty is yours! Send your PayPal to george@tinygrad.org to claim

We should bring `external_test_onnx_backend` out of external and into CI to ensure this doesn't regress

**Comment by geohot** [comment]
BTW, I have three failing tests

```
======================================================================
FAIL: test_resize_upsample_sizes_nearest_axes_2_3_cpu (__main__.OnnxBackendNodeModelTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/batman/.pyenv/versions/3.8.10/lib/python3.8/site-packages/onnx/backend/test/runner/__init__.py", line 296, in device_test_func
    return test_func(*args, device=device, **kwargs)
  File "/home/batman/.pyenv/versions/3.8.10/lib/python3.8/site-packages/onnx/backend/test/runner/__init__.py", line 360, in run
    self.assert_similar_outputs(
  File "/home/batman/.pyenv/versions/3.8.10/lib/python3.8/site-packages/onnx/backend/test/runner/__init__.py", line 210, in assert_similar_outputs
    np.testing.assert_allclose(
  File "/home/batman/.pyenv/versions/3.8.10/lib/python3.8/site-packages/numpy/testing/_private/utils.py", line 1592, in assert_allclose
    assert_array_compare(compare, actual, desired, err_msg=str(err_msg),
  File "/home/batman/.pyenv/versions/3.8.10/lib/python3.8/contextlib.py", line 75, in inner
    return func(*args, **kwds)
  File "/home/batman/.pyenv/versions/3.8.10/lib/python3.8/site-packages/numpy/testing/_private/utils.py", line 862, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Not equal to tolerance rtol=0.001, atol=1e-07

Mismatched elements: 8 / 56 (14.3%)
Max absolute difference: 2.
Max relative difference: 2.
 x: array([[[[1., 1., 1., 1., 2., 2., 2., 2.],
         [1., 1., 1., 1., 2., 2., 2., 2.],
         [1., 1., 1., 1., 2., 2., 2., 2.],...
 y: array([[[[1., 1., 1., 1., 2., 2., 2., 2.],
         [1., 1., 1., 1., 2., 2., 2., 2.],
         [1., 1., 1., 1., 2., 2., 2., 2.],...

======================================================================
FAIL: test_resize_upsample_sizes_nearest_axes_3_2_cpu (__main__.OnnxBackendNodeModelTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/batman/.pyenv/versions/3.8.10/lib/python3.8/site-packages/onnx/backend/test/runner/__init__.py", line 296, in device_test_func
    return test_func(*args, device=device, **kwargs)
  File "/home/batman/.pyenv/versions/3.8.10/lib/python3.8/site-packages/onnx/backend/test/runner/__init__.py", line 360, in run
    self.assert_similar_outputs(
  File "/home/batman/.pyenv/versions/3.8.10/lib/python3.8/site-packages/onnx/backend/test/runner/__init__.py", line 210, in assert_similar_outputs
    np.testing.assert_allclose(
  File "/home/batman/.pyenv/versions/3.8.10/lib/python3.8/site-packages/numpy/testing/_private/utils.py", line 1592, in assert_allclose
    assert_array_compare(compare, actual, desired, err_msg=str(err_msg),
  File "/home/batman/.pyenv/versions/3.8.10/lib/python3.8/contextlib.py", line 75, in inner
    return func(*args, **kwds)
  File "/home/batman/.pyenv/versions/3.8.10/lib/python3.8/site-packages/numpy/testing/_private/utils.py", line 862, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Not equal to tolerance rtol=0.001, atol=1e-07

Mismatched elements: 8 / 56 (14.3%)
Max absolute difference: 2.
Max relative difference: 2.
 x: array([[[[1., 1., 1., 1., 2., 2., 2., 2.],
         [1., 1., 1., 1., 2., 2., 2., 2.],
         [1., 1., 1., 1., 2., 2., 2., 2.],...
 y: array([[[[1., 1., 1., 1., 2., 2., 2., 2.],
         [1., 1., 1., 1., 2., 2., 2., 2.],
         [1., 1., 1., 1., 2., 2., 2., 2.],...

======================================================================
FAIL: test_resize_upsample_sizes_nearest_cpu (__main__.OnnxBackendNodeModelTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/batman/.pyenv/versions/3.8.10/lib/python3.8/site-packages/onnx/backend/test/runner/__init__.py", line 296, in device_test_func
    return test_func(*args, device=device, **kwargs)
  File "/home/batman/.pyenv/versions/3.8.10/lib/python3.8/site-packages/onnx/backend/test/runner/__init__.py", line 360, in run
    self.assert_similar_outputs(
  File "/home/batman/.pyenv/versions/3.8.10/lib/python3.8/site-packages/onnx/backend/test/runner/__init__.py", line 210, in assert_similar_outputs
    np.testing.assert_allclose(
  File "/home/batman/.pyenv/versions/3.8.10/lib/python3.8/site-packages/numpy/testing/_private/utils.py", line 1592, in assert_allclose
    assert_array_compare(compare, actual, desired, err_msg=str(err_msg),
  File "/home/batman/.pyenv/versions/3.8.10/lib/python3.8/contextlib.py", line 75, in inner
    return func(*args, **kwds)
  File "/home/batman/.pyenv/versions/3.8.10/lib/python3.8/site-packages/numpy/testing/_private/utils.py", line 862, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Not equal to tolerance rtol=0.001, atol=1e-07

Mismatched elements: 8 / 56 (14.3%)
Max absolute difference: 2.
Max relative difference: 2.
 x: array([[[[1., 1., 1., 1., 2., 2., 2., 2.],
         [1., 1., 1., 1., 2., 2., 2., 2.],
         [1., 1., 1., 1., 2., 2., 2., 2.],...
 y: array([[[[1., 1., 1., 1., 2., 2., 2., 2.],
         [1., 1., 1., 1., 2., 2., 2., 2.],
         [1., 1., 1., 1., 2., 2., 2., 2.],...

----------------------------------------------------------------------
Ran 2492 tests in 76.107s
```

**Comment by geohot** [comment]
CUDA failures:
```
.sssssssssssss
======================================================================
FAIL: test_constantofshape_int_shape_zero_cpu (__main__.OnnxBackendNodeModelTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/batman/.pyenv/versions/3.8.10/lib/python3.8/site-packages/onnx/backend/test/runner/__init__.py", line 296, in device_test_func
    return test_func(*args, device=device, **kwargs)
  File "/home/batman/.pyenv/versions/3.8.10/lib/python3.8/site-packages/onnx/backend/test/runner/__init__.py", line 359, in run
    outputs = list(prepared_model.run(inputs))
  File "test/external/external_test_onnx_backend.py", line 23, in run
    return tuple(x.numpy() if isinstance(x, Tensor) else [i.numpy() for i in x] if isinstance(x, list) else np.array(x) for x in ret.values())
  File "test/external/external_test_onnx_backend.py", line 23, in <genexpr>
    return tuple(x.numpy() if isinstance(x, Tensor) else [i.numpy() for i in x] if isinstance(x, list) else np.array(x) for x in ret.values())
  File "/home/batman/xx/openpilot/tinygrad/tensor.py", line 103, in numpy
    def numpy(self) -> np.ndarray: return self.lazydata.toCPU()
  File "/home/batman/xx/openpilot/tinygrad/lazy.py", line 195, in toCPU
    realized = self_casted.contiguous().realize().realized
  File "/home/batman/xx/openpilot/tinygrad/lazy.py", line 160, in realize
    self.realized = Device[self.device].exec_ast(self.op, output=self, **self._device_extra_args())
  File "/home/batman/xx/openpilot/tinygrad/ops.py", line 211, in exec_ast
    if k.key not in self.method_cache: self.method_cache[k.key] = get_program()
  File "/home/batman/xx/openpilot/tinygrad/ops.py", line 207, in get_program
    elif not getenv("NOOPT"): k.hand_coded_optimizations()
  File "/home/batman/xx/openpilot/tinygrad/codegen/optimizer.py", line 351, in hand_coded_optimizations
    if self.full_shape[axis] <= 7 and any(st.axis_is_masked(axis) for st in self.sts) and prod(self.full_shape[self.shape_len - self.upcasted:]) * self.full_shape[axis] <= 7 * 7:
  File "/home/batman/xx/openpilot/tinygrad/codegen/optimizer.py", line 351, in <genexpr>
    if self.full_shape[axis] <= 7 and any(st.axis_is_masked(axis) for st in self.sts) and prod(self.full_shape[self.shape_len - self.upcasted:]) * self.full_shape[axis] <= 7 * 7:
  File "/home/batman/xx/openpilot/tinygrad/shape/shapetracker.py", line 198, in axis_is_masked
    _, valid = self.expr_idxs()
  File "/home/batman/xx/openpilot/tinygrad/shape/shapetracker.py", line 188, in expr_idxs
    if idxs is None: idxs = [Variable(f"idx{i}", 0, s-1) for i,s in enumerate(self.shape)]
  File "/home/batman/xx/openpilot/tinygrad/shape/shapetracker.py", line 188, in <listcomp>
    if idxs is None: idxs = [Variable(f"idx{i}", 0, s-1) for i,s in enumerate(self.shape)]
  File "/home/batman/xx/openpilot/tinygrad/shape/symbolic.py", line 148, in __new__
    assert nmin >= 0 and nmin <= nmax
AssertionError

======================================================================
FAIL: test_slice_start_out_of_bounds_cpu (__main__.OnnxBackendNodeModelTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/batman/.pyenv/versions/3.8.10/lib/python3.8/site-packages/onnx/backend/test/runner/__init__.py", line 296, in device_test_func
    return test_func(*args, device=device, **kwargs)
  File "/home/batman/.pyenv/versions/3.8.10/lib/python3.8/site-packages/onnx/backend/test/runner/__init__.py", line 359, in run
    outputs = list(prepared_model.run(inputs))
  File "test/external/external_test_onnx_backend.py", line 23, in run
    return tuple(x.numpy() if isinstance(x, Tensor) else [i.numpy() for i in x] if isinstance(x, list) else np.array(x) for x in ret.values())
  File "test/external/external_test_onnx_backend.py", line 23, in <genexpr>
    return tuple(x.numpy() if isinstance(x, Tensor) else [i.numpy() for i in x] if isinstance(x, list) else np.array(x) for x in ret.values())
  File "/home/batman/xx/openpilot/tinygrad/tensor.py", line 103, in numpy
    def numpy(self) -> np.ndarray: return self.lazydata.toCPU()
  File "/home/batman/xx/openpilot/tinygrad/lazy.py", line 195, in toCPU
    realized = self_casted.contiguous().realize().realized
  File "/home/batman/xx/openpilot/tinygrad/lazy.py", line 160, in realize
    self.realized = Device[self.device].exec_ast(self.op, output=self, **self._device_extra_args())
  File "/home/batman/xx/openpilot/tinygrad/ops.py", line 211, in exec_ast
    if k.key not in self.method_cache: self.method_cache[k.key] = get_program()
  File "/home/batman/xx/openpilot/tinygrad/ops.py", line 207, in get_program
    elif not getenv("NOOPT"): k.hand_coded_optimizations()
  File "/home/batman/xx/openpilot/tinygrad/codegen/optimizer.py", line 160, in hand_coded_optimizations
    self.process()
  File "/home/batman/xx/openpilot/tinygrad/codegen/optimizer.py", line 21, in process
    self.simplify_merge_adjacent()
  File "/home/batman/xx/openpilot/tinygrad/codegen/optimizer.py", line 79, in simplify_merge_adjacent
    for i,x in enumerate(rets): self.sts[i].reshape(tuple([y[0] for y in x]))
  File "/home/batman/xx/openpilot/tinygrad/shape/shapetracker.py", line 244, in reshape
    assert all(is_sym_int(x) and x > 0 for x in new_shape), f"shape must be symbolic ints and can't contain 0 or negative numbers {new_shape}"
AssertionError: shape must be symbolic ints and can't contain 0 or negative numbers (20, 0)

----------------------------------------------------------------------
```
